Add a bitwise NOT operator for integer and Boolean types (CUDA).

xuhdev opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #22326 Delegate Python ~ (invert operator) to Tensor.bitwise_not().
* **#22320 Add a bitwise NOT operator for integer and Boolean types (CUDA).**
* #22283 Add a bitwise NOT operator for integer and Boolean types (CPU).

Related to #20024, #22246, #22262

Differential Revision: [D16183578](https://our.internmc.facebook.com/intern/diff/D16183578)

xuhdev committed:
Add a bitwise NOT operator for integer and Boolean types (CUDA).

Related to #20024, #22246, #22262

xuhdev committed:
Update on "Add a bitwise NOT operator for integer and Boolean types (CUDA)."

Add a bitwise NOT operator for integer and Boolean types (CUDA).

Related to #20024, #22246, #22262

gh-metadata: pytorch pytorch 22320 gh/xuhdev/7/head

xuhdev committed:
Update on "Add a bitwise NOT operator for integer and Boolean types (CUDA)."

Add a bitwise NOT operator for integer and Boolean types (CUDA).

Related to #20024, #22246, #22262

gh-metadata: pytorch pytorch 22320 gh/xuhdev/7/head

xuhdev committed:
Update on "Add a bitwise NOT operator for integer and Boolean types (CUDA)."

Add a bitwise NOT operator for integer and Boolean types (CUDA).

Related to #20024, #22246, #22262

gh-metadata: pytorch pytorch 22320 gh/xuhdev/7/head

xuhdev committed:
Update on "Add a bitwise NOT operator for integer and Boolean types (CUDA)."

Add a bitwise NOT operator for integer and Boolean types (CUDA).

Related to #20024, #22246, #22262

gh-metadata: pytorch pytorch 22320 gh/xuhdev/7/head

xuhdev committed:
Update on "Add a bitwise NOT operator for integer and Boolean types (CUDA)."

Add a bitwise NOT operator for integer and Boolean types (CUDA).

Related to #20024, #22246, #22262

gh-metadata: pytorch pytorch 22320 gh/xuhdev/7/head

xuhdev committed:
Update on "Add a bitwise NOT operator for integer and Boolean types (CUDA)."

Add a bitwise NOT operator for integer and Boolean types (CUDA).

Related to #20024, #22246, #22262

gh-metadata: pytorch pytorch 22320 gh/xuhdev/7/head

xuhdev committed:
Update on "Add a bitwise NOT operator for integer and Boolean types (CUDA)."

Add a bitwise NOT operator for integer and Boolean types (CUDA).

Related to #20024, #22246, #22262

gh-metadata: pytorch pytorch 22320 gh/xuhdev/7/head

colesbury reviewed (changes requested):
This doesn't look like it will work on non-contiguous Tensors. Use TensorIterator here, like you do for the CPU kernel. I think the only difference from the CPU version will be that you should call `gpu_kernel` instead of `cpu_kernel`

xuhdev committed:
Update on "Add a bitwise NOT operator for integer and Boolean types (CUDA)."

Add a bitwise NOT operator for integer and Boolean types (CUDA).

Related to #20024, #22246, #22262

gh-metadata: pytorch pytorch 22320 gh/xuhdev/7/head

xuhdev committed:
Update on "Add a bitwise NOT operator for integer and Boolean types (CUDA)."

Add a bitwise NOT operator for integer and Boolean types (CUDA).

Related to #20024, #22246, #22262

gh-metadata: pytorch pytorch 22320 gh/xuhdev/7/head

xuhdev committed:
Update on "Add a bitwise NOT operator for integer and Boolean types (CUDA)."

Add a bitwise NOT operator for integer and Boolean types (CUDA).

Related to #20024, #22246, #22262

gh-metadata: pytorch pytorch 22320 gh/xuhdev/7/head

xuhdev committed:
Update on "Add a bitwise NOT operator for integer and Boolean types (CUDA)."

Add a bitwise NOT operator for integer and Boolean types (CUDA).

Related to #20024, #22246, #22262

gh-metadata: pytorch pytorch 22320 gh/xuhdev/7/head

colesbury commented on aten/src/ATen/native/cuda/UnaryOpsKernel.cu:
```diff
@@ -1,14 +1,36 @@
+  }
+}
+
+static void bitwise_not_kernel_cuda(TensorIterator& iter) {
```
You can get rid of the static wrapper. It's not really doing anything here.

Previously, when we supported CUDA 8, this pattern was necessary. But now we only have the Windows limitations (i.e. bitwise_not_kernel_cuda_impl can't be static).

colesbury commented on test/test_cuda.py:
```diff
@@ -1088,6 +1088,40 @@ def test_type_conversions_same_gpu(self):
     def test_neg(self):
         _TestTorchMixin._test_neg(self, lambda t: t.cuda())
 
+    def test_bitwise_not(self):
```
I think the CUDA tests should basically be the same as the CPU test. Instead of writing them twice (once in test_torch.py and once in test_cuda.py) you can either:

- follow the pattern of `test_neg` above
- Test on "torch.testing.get_all_device_types()" in test_torch.py like here:
https://github.com/pytorch/pytorch/blob/b76877728a1f06c936a63e7028651561ca3248fc/test/test_torch.py#L3334-L3335

xuhdev replied:
I adopted the test_neg way

colesbury commented on aten/src/ATen/native/cuda/CUDAUnaryOps.cpp:
```diff
@@ -94,4 +97,19 @@ IMPLEMENT_UNARY_OP_PREQUEL(tan)
 IMPLEMENT_UNARY_OP_PREQUEL(tanh)
 IMPLEMENT_UNARY_OP_PREQUEL(trunc)
 
+#define IMPLEMENT_UNARY_OP_VEC_CUDA(op)                          \
```
I think this part can and should be shared between CPU and CUDA. BinaryOps.cpp has examples of this. It might mean that you won't be able to use the existing IMPLEMENT_UNARY_XXX macros.

xuhdev committed:
Update on "Add a bitwise NOT operator for integer and Boolean types (CUDA)."

Add a bitwise NOT operator for integer and Boolean types (CUDA).

Related to #20024, #22246, #22262

gh-metadata: pytorch pytorch 22320 gh/xuhdev/7/head

xuhdev committed:
Update on "Add a bitwise NOT operator for integer and Boolean types (CUDA)."

Add a bitwise NOT operator for integer and Boolean types (CUDA).

Related to #20024, #22246, #22262

gh-metadata: pytorch pytorch 22320 gh/xuhdev/7/head

xuhdev committed:
Update on "Add a bitwise NOT operator for integer and Boolean types (CUDA)."

Add a bitwise NOT operator for integer and Boolean types (CUDA).

Related to #20024, #22246, #22262

gh-metadata: pytorch pytorch 22320 gh/xuhdev/7/head

xuhdev committed:
Update on "Add a bitwise NOT operator for integer and Boolean types (CUDA)."

Add a bitwise NOT operator for integer and Boolean types (CUDA).

Related to #20024, #22246, #22262

gh-metadata: pytorch pytorch 22320 gh/xuhdev/7/head

xuhdev committed:
Update on "Add a bitwise NOT operator for integer and Boolean types (CUDA)."

Add a bitwise NOT operator for integer and Boolean types (CUDA).

Related to #20024, #22246, #22262

gh-metadata: pytorch pytorch 22320 gh/xuhdev/7/head

xuhdev committed:
Update on "Add a bitwise NOT operator for integer and Boolean types (CUDA)."

Add a bitwise NOT operator for integer and Boolean types (CUDA).

Related to #20024, #22246, #22262

gh-metadata: pytorch pytorch 22320 gh/xuhdev/7/head

xuhdev committed:
Update on "Add a bitwise NOT operator for integer and Boolean types (CUDA)."

Add a bitwise NOT operator for integer and Boolean types (CUDA).

Related to #20024, #22246, #22262

gh-metadata: pytorch pytorch 22320 gh/xuhdev/7/head

xuhdev committed:
Update on "Add a bitwise NOT operator for integer and Boolean types (CUDA)."

Add a bitwise NOT operator for integer and Boolean types (CUDA).

Related to #20024, #22246, #22262

gh-metadata: pytorch pytorch 22320 gh/xuhdev/7/head

VitalyFedyunin commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -362,12 +362,11 @@
 - func: bitwise_not(Tensor self, *, Tensor(a!) out) -> Tensor(a!)
   dispatch:
-    CPU: _bitwise_not_out_cpu
+    CPU: bitwise_not_out
```
No need to list them if the dispatch to the same function.

ngimel replied:
There are other places like this, e.g. https://github.com/pytorch/pytorch/blob/d9e15bccb0a127314850567aaf0fea0a6364c3eb/aten/src/ATen/native/native_functions.yaml#L2753-L2754 where CPU and CUDA dispatches go to the same functions - I assumed this is so that sparse tensors error out with a nice message if they call this function? If that's the case, makes sense to leave it here too.

xuhdev committed:
Update on "Add a bitwise NOT operator for integer and Boolean types (CUDA)."

Add a bitwise NOT operator for integer and Boolean types (CUDA).

Related to #20024, #22246, #22262

gh-metadata: pytorch pytorch 22320 gh/xuhdev/7/head
